### ISSUE TITLE
Error building GLOO

### ISSUE THREAD

**Issue by tobyclh** [pytorch/pytorch#4906]
Hello I'm trying to build from source for multinorminal distribution, but I can't get it built.
In contrary, if built with NO_DISTRIBUTED=1 the build passes but of course things like pooling won't work, which is something I need.

- OS: Ubuntu 16.04
- PyTorch version: From master branch
- How you installed PyTorch (conda, pip, source): source 
- CUDA/cuDNN version: 9.1
- GCC version (if compiling from source): 5.4
- Error messages and/or stack traces of the bug
> .../pytorch/torch/lib/gloo/gloo/cuda.h:25:2: error: #error "Expected GLOO_USE_CUDA to be defined"
>  #error "Expected GLOO_USE_CUDA to be defined"
> CMake Error at gloo_cuda_generated_cuda_private.cu.o.cmake:203 (message):
>   Error generating
>   .../pytorch/torch/lib/build/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda_private.cu.o
> 
> 
But make itself seems to be aware of CUDA 9.1

> Install the project...
> -- Install configuration: "Release"
> -- Installing: .../pytorch/torch/lib/tmp_install/lib/libshm.so
> -- Set runtime path of ".../pytorch/torch/lib/tmp_install/lib/libshm.so" to ""
> -- Up-to-date: .../pytorch/torch/lib/tmp_install/include/libshm.h
> -- Installing: ...pytorch/torch/lib/tmp_install/bin/torch_shm_manager
> -- CUDA detected: 9.1
> -- Added CUDA NVCC flags for: sm_30 sm_35 sm_50 sm_52 sm_60 sm_61 sm_70
> -- Found libcuda: /usr/lib/x86_64-linux-gnu/libcuda.so
> -- Found libnvrtc: /usr/local/cuda/lib64/libnvrtc.so
> -- Configuring done
> -- Generating done
> -- Build files have been written to: .../pytorch/torch/lib/build/gloo
>

**Comment by SsnL** [comment]
Could you clarify what you mean by pooling not working?

**Comment by zyoohv** [comment]
hello, I face the same trouble, could you tell me how to solve it ?

**Comment by bmharper** [comment]
I have the same issue. PyTorch 0.4.0, Ubuntu 16.04, Cuda 9.1, CuDNN 7.1.3

**Comment by ezyang** [comment]
I was able to reproduce this issue on 5485f09f1834381262b71a0dff6514b5b5c157eb

**Comment by ezyang** [comment]
However, when I did a hard clean of all of the source directories (including submodules; e.g., `git clean -xfd && git submodule foreach git clean -xfd`) the problem went away. My guess is that if you do a non-CUDA, then do CUDA build, without cleaning in between, that's what causes the problem.

**Comment by ezyang** [comment]
I'm coming back to say, yes, this is exactly what is happening.

**Comment by ezyang** [comment]
This is still happening

**Comment by skyline75489** [comment]
Yeah I can confirm this is still happening on 60eded6c0f34aac79e3adbce3c1d713d5e9bc0b0

**Comment by tyoc213** [comment]
Currently trying to compile latest master... but even that it is included as third party, it can't compile, I hve downloaded https://github.com/facebookincubator/gloo and compiled it without problems...

```
Collecting environment information...
PyTorch version: 1.8.0a0+05c8cd7
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.19.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 3090
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.0
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] torchvision==0.9.0a0+7ffc952
[conda] numpy                     1.20.1                   pypi_0    pypi
[conda] torch                     1.9.0a0+a61a8d0           dev_0    <develop>
[conda] torchvision               0.9.0a0+7ffc952           dev_0    <develop>

```
